### PR TITLE
Introduce —no and modernize AFFIRMATIVE/NEGATIVE

### PR DESCRIPTION
**Pull request by weitzman**
And stop overloading -n|--no-interaction.

### CODE REVIEW

**Comment by greg-1-anderson** [comment]
I don't think we need to handle `yes` and `no` during preflight. They can be ordinary global options, I would expect.

**Comment by weitzman** [comment]
Do you think that we should set runtime.affirmative and runtime.negative configuration during Preflight? Or maybe all we need is Drush:affirmative() and Drush::negative() which calculate the right return value when called?

**Comment by weitzman** [comment]
I just added Drush::affirmative() and Drush::negative() but the global option value is not set properly for some reason. I'm running `sut si --no` and still getting stopped at a prompt. I would expect an immediate cancellation to happen

**Comment by greg-1-anderson** [comment]
Copying global options not handled in preflight is the responsibility of the [GlobalOptionsEventListener](https://github.com/consolidation/Robo/blob/master/src/GlobalOptionsEventListener.php#L44). Anything in the config `options` space will fetch its value from the `$input` option. We have to pre-populate `options` with all of our global options. Currently we do this in drush.yml.

It might be better to not have to name our options twice; however, I don't think that we can tell the difference between global options and local options just by looking at $input.

Maybe we should [have our Application object](https://github.com/drush-ops/drush/blob/master/src/Application.php#L53:L98) initialized the config `options` section. It could just loop over the definition object after initializing it.

**Comment by greg-1-anderson** [comment]
We are also somewhat limited by https://github.com/drush-ops/drush/issues/2956.

File: src/Application.php [hunk 1]
```diff
@@ -26,13 +29,17 @@
...
+class Application extends SymfonyApplication implements LoggerAwareInterface, ConfigAwareInterface
 {
     use LoggerAwareTrait;
+    use ConfigAwareTrait;
 
     /** @var BootstrapManager */
     protected $bootstrapManager;
 
+    /** @var ConfigInterface */
+    protected $config;
```
Comment (greg-1-anderson) [inline src/Application.php]: ConfigAwareTrait provides this

**Comment by greg-1-anderson** [comment]
+1 LGTM

Could remove unnecessary `protected $config;`

**Comment by greg-1-anderson** [comment]
Not sure about the cause of the test failures but this looks good once that is cleared up.

File: src/Application.php [hunk 2]
```diff
@@ -105,8 +105,11 @@ public function configureGlobalOptions()
             );
 
         foreach ($this->getDefinition()->getOptions() as $key => $option) {
-            // The config set here will be populated with resolved values by a later listener.
-            $this->getConfig()->set("options.$key", $option->acceptValue() ? $option->getDefault() : null);
+            // Don't override top level keys set via PreflightArgs::optionsWithValues.
+            if (!$this->getConfig()->has("$key")) {
```
Comment (greg-1-anderson) [inline src/Application.php]: `"options.$key"`?

**Comment by weitzman** [comment]
Just to see what would happen, I commented out the line that was setting config to a default value for the global options. Now we have only one failure, and I'm unclear if its really undesired behavior. @greg-1-anderson - see last commit and last test run. Thanks.

**Comment by weitzman** [comment]
OK, that test fail was due to an unneeded change so I reverted it. This should come back green. Awaiting input from @greg-1-anderson about why this line is needed at all - https://github.com/drush-ops/drush/blob/0a06d83892f2a75f7a39a166f4edfd98acdd2b62/src/Application.php#L111

**Comment by greg-1-anderson** [comment]
Without that line we have:
```
$ ./drush ev 'return \Drush\Drush::config()->get("options");'
array (
  'width' => '89',
)
```

**Comment by greg-1-anderson** [comment]
I'm going to work on making this better in Robo, so that we do not need these extra machinations at all.

**Comment by danepowell** [comment]
Just to be clear, once this is merged we'll have three options for non-interactive calls to Drush:
* `-n|--no-interaction`: assume default values for all prompts
* `--yes`: assume affirmative for all prompts
* `--no`: assume negative for all prompts

Is that correct?

**Comment by greg-1-anderson** [comment]
I removed the `"options.$key"` stuff, which is not necessary with the improved versions of consolidation/config and Robo included in this PR.

**Comment by weitzman** [comment]
@danepowell - no, we cant do that. we dont know if a question is affirmative or negative. So behavior of this PR is:

CORFIRM PROMPT
-n|--no-interaction: proced
--yes: assume default proceed
--no: abort

CHOICE PROMPT
-n|--no-interaction: use default value
--yes: assume default use default value
--no: No effect, AFAICT. user is asked as usual.

**Comment by greg-1-anderson** [comment]
In Drush 8, `--no` caused Drush to abort with an error if a multi-select choice came up.

I'm not sure whether 18eac45 finishes everything here; I'll test a bit more and see. Getting close.

**Comment by greg-1-anderson** [comment]
This seems to be working.  

Continue using default values:
```
$ drush -n si
 [notice] Starting Drupal installation. This takes a while. Consider using the --notify global option.
```

Accept all confirmation prompts:
```
$ drush --yes si

 // You are about to DROP all tables in your 'unish_stage' database. Do you want to      
 // continue?: yes.                                                                      

 [notice] Starting Drupal installation. This takes a while. Consider using the --notify global option.
```

Abort all confirmation prompts:
```
$ drush si --no
                                                                                     
 [WARNING] You are about to DROP all tables in your 'unish_dev' database. Do you want to 
           continue?: no.                                                                
                                                                                         
 [error]  Cancelled. 
```

It appears that a test failure from the master branch was merged in here.

**Comment by greg-1-anderson** [comment]
I think this is all fixed up. Go ahead and merge when ready.

**Comment by greg-1-anderson** [comment]
Oh, actually I should tag a release in Robo before we merge here. Please review and let me know if you're ready for that.

**Pull request merged**